### PR TITLE
Fixed generated call to `ValueType.GetHashCode`

### PR DESCRIPTION
**Pull request by lostmsu**
The bug (https://github.com/m4rs-mt/ILGPU/issues/614) occured when type `T` inside `SpecializedValue<T>` does not implement `GetHashCode`, and instead inherits one from `ValueType`.

Because `ValueType` itself is not a struct, to call its `GetHashCode` the first parameter must be an object reference. Therefore the object of type `T` (which `SpecializedValue<T>` requires to be a `struct`) must be boxed before calling `GetHashCode`.

### CODE REVIEW

File: Src/ILGPU/Backends/IL/ILEmitterExtensions.cs
Comment (MoFtZ) [inline Src/ILGPU/Backends/IL/ILEmitterExtensions.cs:80]: MINOR: Coding convention.
```suggestion
                }
                else
                {
```

File: Src/ILGPU/Backends/IL/ILEmitterExtensions.cs
Comment (MoFtZ) [inline Src/ILGPU/Backends/IL/ILEmitterExtensions.cs:75]: MINOR: This condition can probably be simplified since `SpecializedValue` only accepts `unmanaged` types.
```suggestion
                if (fieldHashCode.DeclaringType != field.FieldType)
```

**Review by MoFtZ — CHANGES_REQUESTED** [review]
Nice work @lostmsu, appears to behave as expected.

Could you please also add a unit test for this condition? Update `Src/ILGPU.Tests/SpecializedKernels.cs` based on the minimal example.

File: Src/ILGPU/Backends/IL/ILEmitterExtensions.cs
Comment (lostmsu) [reply Src/ILGPU/Backends/IL/ILEmitterExtensions.cs:75]: @MoFtZ I was looking forward to it supporting other types as requested in https://github.com/m4rs-mt/ILGPU/issues/415 and https://github.com/m4rs-mt/ILGPU/issues/415

Comment (MoFtZ) [reply Src/ILGPU/Backends/IL/ILEmitterExtensions.cs:75]: We can add support later.

`SpecializedValue` applies the `unmanaged` restriction, and `Action<>` and `Func<>` are reference types, so they cannot be used unless we change `SpecializedValue`, introduce a new type like `SpecalizedFuncValue`, or some other API change.

Also, the original condition would apply boxing if "it is a ValueType and does not match the declaring type". However, wouldn't we want to apply boxing to reference types as well? Easier to deal with it later when we have defined an API, and unit tests to ensure it all works.

Comment (lostmsu) [reply Src/ILGPU/Backends/IL/ILEmitterExtensions.cs:75]: @MoFtZ AFAIK reference type values are "boxes". Which means the version I made does not handle them correctly. They only need `Ldfld` (and not `Ldflda`). If neither of us will be the one to implement specialized lambdas, the person who would will likely stumble upon this error and get puzzled too.

I will implement this code for reference types, but won't add test coverage.

**Comment by lostmsu** [comment]
Added support for reference types.

Added test for struct without explicit `GetHashCode`

File: Src/ILGPU/Backends/IL/ILEmitterExtensions.cs
Comment (lostmsu) [reply Src/ILGPU/Backends/IL/ILEmitterExtensions.cs]: This should be fixed now.

**Comment by lostmsu** [comment]
disabled warning for test struct

**Review by MoFtZ — APPROVED** [review]

**Comment by lostmsu** [comment]
Do you need anything from me to complete this change?

**Comment by MoFtZ** [comment]
@lostmsu I do not have permissions on this repository. Have to wait for @m4rs-mt to review this PR.

**Review by m4rs-mt — APPROVED** [review]
Looks good to me and thanks a lot for your patch 🚀 Would you also mind to agree on the copyright transfer on #598 before we merge this PR?

**Comment by lostmsu** [comment]
@m4rs-mt done

**Pull request merged**
